### ISSUE TITLE
[dtest-release, 6.0] alternator_tests.TesterAlternator.test_alternator_enforce_authorization failed with `StopIteration`

### ISSUE THREAD

**Issue by yaronkaikov** [scylladb/scylladb#19270]
Seen as part of the 6.0.1 release, https://jenkins.scylladb.com/job/scylla-6.0/job/dtest-release/8/

Stacktrace
```
self = <alternator_tests.TesterAlternator object at 0x7fd91b58c410>
request = <FixtureRequest for <Function test_alternator_enforce_authorization>>

    def test_alternator_enforce_authorization(self, request: pytest.FixtureRequest):
        num_of_nodes = self._num_of_nodes_for_test(rf=3)
    
        self.prepare_dynamodb_cluster(
            num_of_nodes=num_of_nodes,
            extra_config={
                "alternator_enforce_authorization": True,
                "authenticator": "org.apache.cassandra.auth.PasswordAuthenticator",
                "authorizer": "org.apache.cassandra.auth.CassandraAuthorizer",
            },
        )
        node1 = self.cluster.nodelist()[0]
    
        # create role, and extract it's salted_hash
        with self.patient_cql_connection(node1, user="cassandra", password="cassandra") as cql_session:
            cql_session.execute("CREATE ROLE alternator WITH PASSWORD = 'password'")
            for ks in ("system", "system_auth_v2", "system_auth"):
                try:
                    response = cql_session.execute(f"SELECT salted_hash FROM {ks}.roles WHERE role=%s;", ("alternator",))
                except InvalidRequest:
                    continue
>               self.salted_hash = next(iter(response)).salted_hash

alternator_tests.py:1356: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
cassandra/cluster.py:5365: in cassandra.cluster.ResultSet.next
    ???
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

>   ???
E   StopIteration

cassandra/cluster.py:5360: StopIteration
```

**Comment by yaronkaikov** [comment]
Maybe related to https://github.com/scylladb/scylla-dtest/commit/aca7ef9a767c7b89743a36a73609df4fb607eb38 or maybe missing a backport? @nyh I assume you are the one to ask :-)

**Comment by nyh** [comment]
Yes, the failing code is in @fruch's dtest patch which I guess was backported to 6.0.

The failing code looks correct to me, and should have worked regardless of what code exactly went into Scylla (it tries three keyspace names, "system", "system_auth_v2", "system_auth", to work with three versions of Scylla).
Does it seem to fail every time on this 6.0 branch, or was it a fluke?

@nuivall is it possible that some versions of Scylla had one of these three keyspaces with the "roles" table in it but it is **empty** (not storing the actual roles)? Because @fruch's test code assumes that either the table ks.roles doesn't exist at all, or it exists and contains the "alternator" role. A mid-state (the table exists but doesn't get updated) will confuse it.

**Comment by nuivall** [comment]
> Yes, the failing code is in @fruch's dtest patch which I guess was backported to 6.0.
> 
> The failing code looks correct to me, and should have worked regardless of what code exactly went into Scylla (it tries three keyspace names, "system", "system_auth_v2", "system_auth", to work with three versions of Scylla). Does it seem to fail every time on this 6.0 branch, or was it a fluke?
> 
> @nuivall is it possible that some versions of Scylla had one of these three keyspaces with the "roles" table in it but it is **empty** (not storing the actual roles)? Because @fruch's test code assumes that either the table ks.roles doesn't exist at all, or it exists and contains the "alternator" role. A mid-state (the table exists but doesn't get updated) will confuse it.

Yes it is possible, due to https://github.com/scylladb/scylladb/issues/10481 (and there is even more split wrt. password https://github.com/scylladb/scylladb/issues/10481#issuecomment-2144594369)

**Comment by nyh** [comment]
> Yes it is possible, due to #10481 (and there is even more split wrt. password [#10481 (comment)](https://github.com/scylladb/scylladb/issues/10481#issuecomment-2144594369))

I'm not sure how this can be the same problem. Can calling CREATE ROLE too early somehow result in the roles table being created in the "wrong" keyspace, or being created in the right one but being left empty?

I have another guess what might be the problem (@fruch, what do you think?): Is it possible that a read of the roles table with CL=ONE (which is the default for reads in the Python driver, IIRC!) from a different coordinator where the CREATE ROLE was done can return old data? If this is the case, changing the roles read in the test to CL=ALL will fix this problem.

**Comment by nuivall** [comment]
ok, I see we are not using the default role (btw, we should as I think test.py does now). so you need a read barrier before select, in auth-v2 writes are strongly consistent but reads eventually.

**Comment by nuivall** [comment]
example from test.py:

```
async def read_barrier(cql: Session, host: Host):
    """To issue a read barrier it is sufficient to attempt dropping a
    non-existing table. We need to use `if exists`, otherwise the statement
    would fail on prepare/validate step which happens before a read barrier is
    performed.
    """
    await cql.run_async("drop table if exists nosuchkeyspace.nosuchtable", host = host)

```

```
    # auth reads are eventually consistent so we need to sync all nodes
    await asyncio.gather(*(read_barrier(cql, host) for host in hosts))
```

btw we could reuse test.py code as duplication effort as we see is huge

**Comment by fruch** [comment]
> ok, I see we are not using the default role (btw, we should as I think test.py does now). so you need a read barrier before select, in auth-v2 writes are strongly consistent but reads eventually.

@temichus, once this procedure is documented, please see dtest/SCT follows it.

**Comment by nuivall** [comment]
There is no documentation for the SELECT part because that's not users do or should do. 

We don't specify whether changes are immediate or not (so it's wrong to assume anything), we document it as a manual procedure. Maybe we can write somewhere that the change is not immediate, but unsure  where, as the manual procedure says already to "verify" that modified auth settings display as they should (maybe there? but it's anyway easy to miss)

**Comment by nyh** [comment]
> ok, I see we are not using the default role (btw, we should as I think test.py does now). so you need a read barrier before select, in auth-v2 writes are strongly consistent but reads eventually.

@fruch thought the "create a new role" more faithfully represents what a real user would do, and he does have a point - we just realized that what well the user - to do a CREATE ROLE and then immediately read the newly created hash key - might not actually work without a ??? (CL=ALL? "read barrier"?). This test might help us figure out what this "???" is and then we should probably document it.

**Comment by fruch** [comment]
> There is no documentation for the SELECT part because that's not users do or should do.

so what should alternator user that want to use this feature of enforcing auth should be doing ? how can they setup correctly their code to use auth ? 

> 
> We don't specify whether changes are immediate or not (so it's wrong to assume anything), we document it as a manual procedure. Maybe we can write somewhere that the change is not immediate, but unsure where, as the manual procedure says already to "verify" that modified auth settings display as they should (maybe there? but it's anyway easy to miss)

one sentence ago you said there's no documentation, and now you that there is. but it's only for manual process, i.e. when use would be doing it slow ? :)

**Comment by nuivall** [comment]
> > There is no documentation for the SELECT part because that's not users do or should do.
> 
> so what should alternator user that want to use this feature of enforcing auth should be doing ? how can they setup correctly their code to use auth ?
> 

waiting for raft propagation and apply, as far as I know it is unspecified how long it can take (usually seconds?), but maybe someone can provide bounds

> > We don't specify whether changes are immediate or not (so it's wrong to assume anything), we document it as a manual procedure. Maybe we can write somewhere that the change is not immediate, but unsure where, as the manual procedure says already to "verify" that modified auth settings display as they should (maybe there? but it's anyway easy to miss)
> 
> one sentence ago you said there's no documentation, and now you that there is. but it's only for manual process, i.e. when use would be doing it slow ? :)

There is no documentation for read barrier and select part, we don't expect user to do this.

**Comment by mykaul** [comment]
It's unclear to me what's the next step here and who owns it.

**Comment by fruch** [comment]
This job is using

--force-gossip-topology-changes

So it's missing backport (at the time)

https://github.com/scylladb/scylla-dtest/pull/4411

As for ownership, this is issue related to auth_v2, so ownership of those issues should be clear...

**Comment by ptrsmrn** [comment]
> Does it seem to fail every time on this 6.0 branch, or was it a fluke?

@nyh From what I see, the latest, 8th, job was run on Jun 13, 2024 and the 7th job was run on Jun 6, 2024. @fruch's [patch](https://github.com/scylladb/scylla-dtest/commit/aca7ef9a767c7b89743a36a73609df4fb607eb38) was merged on Jun 9th, so in between these runs, which may lead to 8th run failing the test. BTW. 7th job was successful, and only 1 testcase (the one this issue is about) failed in the 8th job

**Comment by ptrsmrn** [comment]
This test may be just flaky, because it also failed [on master](https://jenkins.scylladb.com/view/master/job/scylla-master/job/dtest-release/604/testReport/alternator_tests/TesterAlternator/Run_Dtest_Parallel_Cloud_Machines___FullDtest___full_split000___test_alternator_enforce_authorization/history/) (pics attached, because the page is loading slowly) [on the same error](https://jenkins.scylladb.com/view/master/job/scylla-master/job/dtest-release/592/testReport/junit/alternator_tests/TesterAlternator/Run_Dtest_Parallel_Cloud_Machines___FullDtest___full_split000___test_alternator_enforce_authorization/).
![image](https://github.com/scylladb/scylladb/assets/124208650/4f5b7312-524d-4722-94bf-eaea131e5ca1)

**Comment by ptrsmrn** [comment]
So the last failure for master was observed in job no. 592 and since job no. 593 the testcase is passing. The [difference between the jobs 592 & 593](https://jenkins.scylladb.com/view/master/job/scylla-master/job/dtest-release/593/) show [the testcase in questions has been modified](https://github.com/scylladb/scylla-dtest/commit/6ae7cf97d504b940b1fbc4d04d918bc3eda90dde#diff-0).

**Comment by ptrsmrn** [comment]
> This job is using
> 
> --force-gossip-topology-changes
> 
> So it's missing backport (at the time)
> 
> [scylladb/scylla-dtest#4411](https://github.com/scylladb/scylla-dtest/pull/4411)

@fruch is it enough to re-run the job to see if the error is gone? Can you do it? If not, whom should I ask?

> As for ownership, this is issue related to auth_v2, so ownership of those issues should be clear...

@fruch Do you mean that the core-frontend team should do it, since we authored auth v2, or you should, since you wrote the test and the fix for it? It's not clear to me. Better to get it correctly for similar future events.

**Comment by nyh** [comment]
I'm trying to understand the situation here.

Reading above, it seems to me (but please correct me if I'm wrong) that the conclusion was that @fruch 's new test code assumed that it can create a role and immediately read its hash key, but it turns out that one needs to wait for a "barrier" before reading from the roles table. If this is the case, I think that:
1. The core-frontend team (@ptrsmrn) should explain to @fruch and possibly also in the documentation, how exactly one "waits for barrier". If today it's not possible, and they recommend a "sleep 60" or anything else, please say so.
2. @fruch should use whatever incantation was recommended above.

All that being said, I still don't understand why we are seeing this problem just in the 6.0 branch. It makes me worried we are missing something.

**Comment by fruch** [comment]
> I'm trying to understand the situation here.
> 
> Reading above, it seems to me (but please correct me if I'm wrong) that the conclusion was that @fruch 's new test code assumed that it can create a role and immediately read its hash key, but it turns out that one needs to wait for a "barrier" before reading from the roles table. If this is the case, I think that:
> 1. The core-frontend team (@ptrsmrn) should explain to @fruch and possibly also in the documentation, how exactly one "waits for barrier". If today it's not possible, and they recommend a "sleep 60" or anything else, please say so.
> 2. @fruch should use whatever incantation was recommended above.
> 
> All that being said, I still don't understand why we are seeing this problem just in the 6.0 branch. It makes me worried we are missing something.

The reason the test was failing in 6.0 was cause the fallback to the next keyspace wasn't working in the force gossip case.
That was fixed on master, but the backport took some time to attend to and reach 6.0

I don't know if this read barrier is needed or not, and if one thinks it is needed, he should be document and fix the tests.

**Comment by nyh** [comment]
> The reason the test was failing in 6.0 was cause the fallback to the next keyspace wasn't working in the force gossip case. That was fixed on master, but the backport took some time to attend to and reach 6.0

@fruch, has it reached 6.0? If so, should this issue be closed?
